Index: bbp-admin/bbp-admin.php
===================================================================
--- bbp-admin/bbp-admin.php	(revision 2779)
+++ bbp-admin/bbp-admin.php	(working copy)
@@ -1084,7 +1084,7 @@
 	 *
 	 * @param array $actions Actions
 	 * @param array $topic Topic object
-	 * @uses the_content() To output topic content
+	 * @uses bbp_topic_content() To output topic content
 	 * @uses bbp_get_topic_permalink() To get the topic link
 	 * @uses bbp_get_topic_title() To get the topic title
 	 * @uses current_user_can() To check if the current user can edit or
@@ -1106,7 +1106,7 @@
 		if ( $bbp->topic_id == $topic->post_type ) {
 			unset( $actions['inline hide-if-no-js'] );

-			the_content();
+			bbp_topic_content();

 			// Show view link if it's not set, the topic is trashed and the user can view trashed topics
 			if ( empty( $actions['view'] ) && 'trash' == $topic->post_status && current_user_can( 'view_trash' ) )
@@ -1416,7 +1416,7 @@
 	 *
 	 * @param array $actions Actions
 	 * @param array $reply Reply object
-	 * @uses the_content() To output reply content
+	 * @uses bbp_reply_content() To output reply content
 	 * @uses bbp_get_reply_permalink() To get the reply link
 	 * @uses bbp_get_reply_title() To get the reply title
 	 * @uses current_user_can() To check if the current user can edit or
@@ -1439,7 +1439,7 @@
 			if ( empty( $actions['view'] ) && 'trash' == $reply->post_status && current_user_can( 'view_trash' ) )
 				$actions['view'] = '<a href="' . bbp_get_reply_permalink( $reply->ID ) . '" title="' . esc_attr( sprintf( __( 'View &#8220;%s&#8221;', 'bbpress' ), bbp_get_reply_title( $reply->ID ) ) ) . '" rel="permalink">' . __( 'View', 'bbpress' ) . '</a>';

-			the_content();
+			bbp_reply_content();

 			// Only show the actions if the user is capable of viewing them
 			if ( current_user_can( 'moderate', $reply->ID ) ) {
Index: bbp-includes/bbp-functions.php
===================================================================
--- bbp-includes/bbp-functions.php	(revision 2779)
+++ bbp-includes/bbp-functions.php	(working copy)
@@ -258,12 +258,12 @@
  * @uses bbp_set_current_anonymous_user_data() To set the anonymous user
  *                                                cookies
  * @uses is_wp_error() To check if the value retrieved is a {@link WP_Error}
+ * @uses remove_filter() To remove 'wp_filter_kses' filters if needed
  * @uses esc_attr() For sanitization
  * @uses bbp_check_for_flood() To check for flooding
  * @uses bbp_check_for_duplicate() To check for duplicates
- * @uses author_can() To check if the author of the reply can post unfiltered
- *                     html or not
- * @uses wp_filter_post_kses() To filter the post content
+ * @uses apply_filters() Calls 'bbp_new_reply_pre_title' with the title
+ * @uses apply_filters() Calls 'bbp_new_reply_pre_content' with the content
  * @uses wp_set_post_terms() To set the topic tags
  * @uses bbPress::errors::get_error_codes() To get the {@link WP_Error} errors
  * @uses wp_insert_post() To insert the reply
@@ -299,24 +299,34 @@
 				bbp_set_current_anonymous_user_data( $anonymous_data );
 		}

+		// Handle Topic ID to append reply to
+		if ( empty( $_POST['bbp_topic_id'] ) || !$topic_id = $_POST['bbp_topic_id'] )
+			$bbp->errors->add( 'bbp_reply_topic_id', __( '<strong>ERROR</strong>: Topic ID is missing.', 'bbpress' ) );
+
+		// Handle Forum ID to adjust counts of
+		if ( empty( $_POST['bbp_forum_id'] ) || !$forum_id = $_POST['bbp_forum_id'] )
+			$bbp->errors->add( 'bbp_reply_forum_id', __( '<strong>ERROR</strong>: Forum ID is missing.', 'bbpress' ) );
+
+		// Remove wp_filter_kses filters from title and content for capable users and if the nonce is verified
+		if ( current_user_can( 'unfiltered_html' ) && !empty( $_POST['_bbp_unfiltered_html_reply'] ) && wp_create_nonce( 'bbp-unfiltered-html-reply_' . $topic_id ) == $_POST['_bbp_unfiltered_html_reply'] ) {
+			remove_filter( 'bbp_new_reply_pre_title',   'wp_filter_kses' );
+			remove_filter( 'bbp_new_reply_pre_content', 'wp_filter_kses' );
+		}
+
 		// Handle Title (optional for replies)
 		if ( !empty( $_POST['bbp_reply_title'] ) )
 			$reply_title = esc_attr( strip_tags( $_POST['bbp_reply_title'] ) );

+		$reply_title = apply_filters( 'bbp_new_reply_pre_title', $reply_title );
+
 		// Handle Content
-		if ( empty( $_POST['bbp_reply_content'] ) || !$reply_content = ( !bbp_is_anonymous() && author_can( $reply_author, 'unfiltered_html' ) ) ? $_POST['bbp_reply_content'] : wp_filter_post_kses( $_POST['bbp_reply_content'] ) ) {
+		if ( empty( $_POST['bbp_reply_content'] ) || !$reply_content = $_POST['bbp_reply_content'] ) {
 			$bbp->errors->add( 'bbp_reply_content', __( '<strong>ERROR</strong>: Your reply cannot be empty.', 'bbpress' ) );
 			$reply_content = '';
 		}

-		// Handle Topic ID to append reply to
-		if ( empty( $_POST['bbp_topic_id'] ) || !$topic_id = $_POST['bbp_topic_id'] )
-			$bbp->errors->add( 'bbp_reply_topic_id', __( '<strong>ERROR</strong>: Topic ID is missing.', 'bbpress' ) );
+		$reply_content = apply_filters( 'bbp_new_reply_pre_content', $reply_content );

-		// Handle Forum ID to adjust counts of
-		if ( empty( $_POST['bbp_forum_id'] ) || !$forum_id = $_POST['bbp_forum_id'] )
-			$bbp->errors->add( 'bbp_reply_forum_id', __( '<strong>ERROR</strong>: Forum ID is missing.', 'bbpress' ) );
-
 		// Check for flood
 		if ( !bbp_check_for_flood( $anonymous_data, $reply_author ) )
 			$bbp->errors->add( 'bbp_reply_flood', __( '<strong>ERROR</strong>: Slow down; you move too fast.', 'bbpress' ) );
@@ -326,8 +336,7 @@
 			$bbp->errors->add( 'bbp_reply_duplicate', __( '<strong>ERROR</strong>: Duplicate reply detected; it looks as though you&#8217;ve already said that!', 'bbpress' ) );

 		// Handle Tags
-		if ( !empty( $_POST['bbp_topic_tags'] ) ) {
-			$tags = $_POST['bbp_topic_tags'];
+		if ( !empty( $_POST['bbp_topic_tags'] ) && $tags = esc_attr( strip_tags( $_POST['bbp_topic_tags'] ) ) ) {
 			$tags = wp_set_post_terms( $topic_id, $tags, $bbp->topic_tag_id, true );

 			if ( is_wp_error( $tags ) || false == $tags )
@@ -381,10 +390,12 @@
  * @uses current_user_can() To check if the current user can edit that reply
  * @uses bbp_filter_anonymous_post_data() To filter anonymous data
  * @uses is_wp_error() To check if the value retrieved is a {@link WP_Error}
+ * @uses remove_filter() To remove 'wp_filter_kses' filters if needed
  * @uses esc_attr() For sanitization
- * @uses author_can() To check if the author of the reply can post unfiltered
- *                     html or not
- * @uses wp_filter_post_kses() To filter the post content
+ * @uses apply_filters() Calls 'bbp_edit_reply_pre_title' with the title and
+ *                       reply id
+ * @uses apply_filters() Calls 'bbp_edit_reply_pre_content' with the content
+ *                        reply id
  * @uses wp_set_post_terms() To set the topic tags
  * @uses bbPress::errors::get_error_codes() To get the {@link WP_Error} errors
  * @uses wp_update_post() To update the reply
@@ -424,13 +435,22 @@

 		}

+		// Remove wp_filter_kses filters from title and content for capable users and if the nonce is verified
+		if ( current_user_can( 'unfiltered_html' ) && !empty( $_POST['_bbp_unfiltered_html_reply'] ) && wp_create_nonce( 'bbp-unfiltered-html-reply_' . $reply_id ) == $_POST['_bbp_unfiltered_html_reply'] ) {
+			remove_filter( 'bbp_edit_reply_pre_title',   'wp_filter_kses' );
+			remove_filter( 'bbp_edit_reply_pre_content', 'wp_filter_kses' );
+		}
+
 		// Handle Title (optional for replies)
 		$reply_title = !empty( $_POST['bbp_reply_title'] ) ? esc_attr( strip_tags( $_POST['bbp_reply_title'] ) ) : $reply_title = $reply->post_title;
+		$reply_title = apply_filters( 'bbp_edit_reply_pre_title', $reply_title, $reply_id );

 		// Handle Content
-		if ( empty( $_POST['bbp_reply_content'] ) || !$reply_content = ( !bbp_is_reply_anonymous( $reply_id ) && author_can( $reply->post_author, 'unfiltered_html' ) ) ? $_POST['bbp_reply_content'] : wp_filter_post_kses( $_POST['bbp_reply_content'] ) )
+		if ( empty( $_POST['bbp_reply_content'] ) || !$reply_content = $_POST['bbp_reply_content'] )
 			$bbp->errors->add( 'bbp_edit_reply_content', __( '<strong>ERROR</strong>: Your reply cannot be empty.', 'bbpress' ) );

+		$reply_content = apply_filters( 'bbp_edit_reply_pre_content', $reply_content, $reply_id );
+
 		// Handle insertion into posts table
 		if ( !is_wp_error( $bbp->errors ) || !$bbp->errors->get_error_codes() ) {

@@ -562,14 +582,14 @@
  * @uses bbp_set_current_anonymous_user_data() To set the anonymous user cookies
  * @uses is_wp_error() To check if the value retrieved is a {@link WP_Error}
  * @uses esc_attr() For sanitization
- * @uses author_can() To check if the author of the reply can post unfiltered
- *                     html or not
  * @uses bbp_is_forum_category() To check if the forum is a category
  * @uses bbp_is_forum_closed() To check if the forum is closed
  * @uses bbp_is_forum_private() To check if the forum is private
  * @uses bbp_check_for_flood() To check for flooding
  * @uses bbp_check_for_duplicate() To check for duplicates
- * @uses wp_filter_post_kses() To filter the post content
+ * @uses remove_filter() To remove 'wp_filter_kses' filters if needed
+ * @uses apply_filters() Calls 'bbp_new_topic_pre_title' with the content
+ * @uses apply_filters() Calls 'bbp_new_topic_pre_content' with the content
  * @uses bbPress::errors::get_error_codes() To get the {@link WP_Error} errors
  * @uses wp_insert_post() To insert the topic
  * @uses do_action() Calls 'bbp_new_topic' with the topic id, forum id,
@@ -604,14 +624,24 @@
 				bbp_set_current_anonymous_user_data( $anonymous_data );
 		}

+		// Remove wp_filter_kses filters from title and content for capable users and if the nonce is verified
+		if ( current_user_can( 'unfiltered_html' ) && !empty( $_POST['_bbp_unfiltered_html_topic'] ) && wp_create_nonce( 'bbp-unfiltered-html-topic_new' ) == $_POST['_bbp_unfiltered_html_topic'] ) {
+			remove_filter( 'bbp_new_topic_pre_title',   'wp_filter_kses' );
+			remove_filter( 'bbp_new_topic_pre_content', 'wp_filter_kses' );
+		}
+
 		// Handle Title
 		if ( empty( $_POST['bbp_topic_title'] ) || !$topic_title = esc_attr( strip_tags( $_POST['bbp_topic_title'] ) ) )
 			$bbp->errors->add( 'bbp_topic_title', __( '<strong>ERROR</strong>: Your topic needs a title.', 'bbpress' ) );

+		$topic_title = apply_filters( 'bbp_new_topic_pre_title', $topic_title );
+
 		// Handle Content
-		if ( empty( $_POST['bbp_topic_content'] ) || !$topic_content = ( !bbp_is_anonymous() && author_can( $topic_author, 'unfiltered_html' ) ) ? $_POST['bbp_topic_content'] : wp_filter_post_kses( $_POST['bbp_topic_content'] ) )
+		if ( empty( $_POST['bbp_topic_content'] ) || !$topic_content = $_POST['bbp_topic_content'] )
 			$bbp->errors->add( 'bbp_topic_content', __( '<strong>ERROR</strong>: Your topic needs some content.', 'bbpress' ) );

+		$topic_content = apply_filters( 'bbp_new_topic_pre_content', $topic_content );
+
 		// Handle Forum id to append topic to
 		if ( empty( $_POST['bbp_forum_id'] ) || !$forum_id = $_POST['bbp_forum_id'] ) {
 			$bbp->errors->add( 'bbp_topic_forum_id', __( '<strong>ERROR</strong>: Forum ID is missing.', 'bbpress' ) );
@@ -637,7 +667,7 @@
 		// Handle Tags
 		if ( !empty( $_POST['bbp_topic_tags'] ) ) {
 			// Escape tag input
-			$terms = esc_html( $_POST['bbp_topic_tags'] );
+			$terms = esc_attr( strip_tags( $_POST['bbp_topic_tags'] ) );

 			// Explode by comma
 			if ( strstr( $terms, ',' ) )
@@ -700,12 +730,14 @@
  * @uses bbp_filter_anonymous_post_data() To filter anonymous data
  * @uses is_wp_error() To check if the value retrieved is a {@link WP_Error}
  * @uses esc_attr() For sanitization
- * @uses author_can() To check if the author of the reply can post unfiltered
- *                     html or not
  * @uses bbp_is_forum_category() To check if the forum is a category
  * @uses bbp_is_forum_closed() To check if the forum is closed
  * @uses bbp_is_forum_private() To check if the forum is private
- * @uses wp_filter_post_kses() To filter the post content
+ * @uses remove_filter() To remove 'wp_filter_kses' filters if needed
+ * @uses apply_filters() Calls 'bbp_edit_topic_pre_title' with the title and
+ *                        topic id
+ * @uses apply_filters() Calls 'bbp_edit_topic_pre_content' with the content
+ *                        and topic id
  * @uses bbPress::errors::get_error_codes() To get the {@link WP_Error} errors
  * @uses wp_update_post() To update the topic
  * @uses do_action() Calls 'bbp_edit_topic' with the topic id, forum id,
@@ -743,6 +775,12 @@
 			}
 		}

+		// Remove wp_filter_kses filters from title and content for capable users and if the nonce is verified
+		if ( current_user_can( 'unfiltered_html' ) && !empty( $_POST['_bbp_unfiltered_html_topic'] ) && wp_create_nonce( 'bbp-unfiltered-html-topic_' . $topic_id ) == $_POST['_bbp_unfiltered_html_topic'] ) {
+			remove_filter( 'bbp_edit_topic_pre_title',   'wp_filter_kses' );
+			remove_filter( 'bbp_edit_topic_pre_content', 'wp_filter_kses' );
+		}
+
 		// Handle Forum id to append topic to
 		if ( empty( $_POST['bbp_forum_id'] ) || !$forum_id = $_POST['bbp_forum_id'] ) {
 			$bbp->errors->add( 'bbp_topic_forum_id', __( '<strong>ERROR</strong>: Forum ID is missing.', 'bbpress' ) );
@@ -761,10 +799,14 @@
 		if ( empty( $_POST['bbp_topic_title'] ) || !$topic_title = esc_attr( strip_tags( $_POST['bbp_topic_title'] ) ) )
 			$bbp->errors->add( 'bbp_edit_topic_title', __( '<strong>ERROR</strong>: Your topic needs a title.', 'bbpress' ) );

+		$topic_title = apply_filters( 'bbp_edit_topic_pre_title', $topic_title, $topic_id );
+
 		// Handle Content
-		if ( empty( $_POST['bbp_topic_content'] ) || !$topic_content = ( !bbp_is_topic_anonymous( $topic_id ) && author_can( $topic->post_author, 'unfiltered_html' ) ) ? $_POST['bbp_topic_content'] : wp_filter_post_kses( $_POST['bbp_topic_content'] ) )
+		if ( empty( $_POST['bbp_topic_content'] ) || !$topic_content = $_POST['bbp_topic_content'] )
 			$bbp->errors->add( 'bbp_edit_topic_content', __( '<strong>ERROR</strong>: Your topic cannot be empty.', 'bbpress' ) );

+		$topic_content = apply_filters( 'bbp_edit_topic_pre_content', $topic_content, $topic_id );
+
 		// Handle insertion into posts table
 		if ( !is_wp_error( $bbp->errors ) || !$bbp->errors->get_error_codes() ) {

Index: bbp-includes/bbp-general-template.php
===================================================================
--- bbp-includes/bbp-general-template.php	(revision 2779)
+++ bbp-includes/bbp-general-template.php	(working copy)
@@ -459,6 +459,13 @@
 		<input type="hidden" name="action"       id="bbp_post_action" value="bbp-edit-topic" />
 		<input type="hidden" name="bbp_topic_id" id="bbp_topic_id"    value="<?php bbp_topic_id(); ?>" />

+		<?php
+
+		if ( current_user_can( 'unfiltered_html' ) )
+			wp_nonce_field( 'bbp-unfiltered-html-topic_' . bbp_get_topic_id(), '_bbp_unfiltered_html_topic' );
+
+		?>
+
 		<?php wp_nonce_field( 'bbp-edit-topic_' . bbp_get_topic_id() );

 	else :
@@ -471,6 +478,13 @@

 		<input type="hidden" name="action" id="bbp_post_action" value="bbp-new-topic" />

+		<?php
+
+		if ( current_user_can( 'unfiltered_html' ) )
+			wp_nonce_field( 'bbp-unfiltered-html-topic_new', '_bbp_unfiltered_html_topic' );
+
+		?>
+
 		<?php wp_nonce_field( 'bbp-new-topic' );

 	endif;
@@ -494,6 +508,13 @@
 		<input type="hidden" name="action"       id="bbp_post_action" value="bbp-edit-reply" />
 		<input type="hidden" name="bbp_reply_id" id="bbp_reply_id"    value="<?php bbp_reply_id(); ?>" />

+		<?php
+
+		if ( current_user_can( 'unfiltered_html' ) )
+			wp_nonce_field( 'bbp-unfiltered-html-reply_' . bbp_get_reply_id(), '_bbp_unfiltered_html_reply' );
+
+		?>
+
 		<?php wp_nonce_field( 'bbp-edit-reply_' . bbp_get_reply_id() );

 	} else {
@@ -505,6 +526,13 @@
 		<input type="hidden" name="bbp_topic_id"    id="bbp_topic_id"    value="<?php bbp_topic_id(); ?>" />
 		<input type="hidden" name="action"          id="bbp_post_action" value="bbp-new-reply" />

+		<?php
+
+		if ( current_user_can( 'unfiltered_html' ) )
+			wp_nonce_field( 'bbp-unfiltered-html-reply_' . bbp_get_topic_id(), '_bbp_unfiltered_html_reply' );
+
+		?>
+
 		<?php wp_nonce_field( 'bbp-new-reply' );
 	}
 }
@@ -684,6 +712,22 @@
 		return apply_filters( 'bbp_get_breadcrumb', $trail . get_the_title() );
 	}

+/**
+ * Display all of the allowed tags in HTML format with attributes.
+ *
+ * This is useful for displaying in the post area, which elements and
+ * attributes are supported. As well as any plugins which want to display it.
+ *
+ * @since bbPress (r2780)
+ *
+ * @uses allowed_tags() To get the allowed tags
+ * @uses apply_filters() Calls 'bbp_allowed_tags' with the tags
+ * @return string HTML allowed tags entity encoded.
+ */
+function bbp_allowed_tags() {
+	return apply_filters( 'bbp_allowed_tags', allowed_tags() );
+}
+
 /** Start Query Functions *****************************************************/

 /**
Index: bbp-includes/bbp-hooks.php
===================================================================
--- bbp-includes/bbp-hooks.php	(revision 2779)
+++ bbp-includes/bbp-hooks.php	(working copy)
@@ -197,10 +197,56 @@

 /** FILTERS *******************************************************************/

+// wp_filter_kses on new/edit topic/reply title
+add_filter( 'bbp_new_reply_pre_title',  'wp_filter_kses' );
+add_filter( 'bbp_new_topic_pre_title',  'wp_filter_kses' );
+add_filter( 'bbp_edit_reply_pre_title', 'wp_filter_kses' );
+add_filter( 'bbp_edit_topic_pre_title', 'wp_filter_kses' );
+
+// balanceTags, wp_filter_kses and wp_rel_nofollow on new/edit topic/reply text
+add_filter( 'bbp_new_reply_pre_content',  'balanceTags'     );
+add_filter( 'bbp_new_reply_pre_content',  'wp_rel_nofollow' );
+add_filter( 'bbp_new_reply_pre_content',  'wp_filter_kses'  );
+add_filter( 'bbp_new_topic_pre_content',  'balanceTags'     );
+add_filter( 'bbp_new_topic_pre_content',  'wp_rel_nofollow' );
+add_filter( 'bbp_new_topic_pre_content',  'wp_filter_kses'  );
+add_filter( 'bbp_edit_reply_pre_content', 'balanceTags'     );
+add_filter( 'bbp_edit_reply_pre_content', 'wp_rel_nofollow' );
+add_filter( 'bbp_edit_reply_pre_content', 'wp_filter_kses'  );
+add_filter( 'bbp_edit_topic_pre_content', 'balanceTags'     );
+add_filter( 'bbp_edit_topic_pre_content', 'wp_rel_nofollow' );
+add_filter( 'bbp_edit_topic_pre_content', 'wp_filter_kses'  );
+
 // Add number format filter to functions requiring numeric output
 add_filter( 'bbp_get_forum_topic_count',       'bbp_number_format' );
 add_filter( 'bbp_get_forum_topic_reply_count', 'bbp_number_format' );

+// Run wp_kses_data on topic/reply content in admin section
+if ( is_admin() ) {
+	add_filter( 'bbp_get_reply_content', 'wp_kses_data' );
+	add_filter( 'bbp_get_topic_content', 'wp_kses_data' );
+}
+
+// Run filters on reply content
+add_filter( 'bbp_get_reply_content', 'capital_P_dangit'                          );
+add_filter( 'bbp_get_reply_content', 'bbp_reply_content_append_revisions',  1, 2 );
+add_filter( 'bbp_get_reply_content', 'wptexturize',                         3    );
+add_filter( 'bbp_get_reply_content', 'convert_chars',                       5    );
+add_filter( 'bbp_get_reply_content', 'make_clickable',                      9    );
+add_filter( 'bbp_get_reply_content', 'force_balance_tags',                 25    );
+add_filter( 'bbp_get_reply_content', 'convert_smilies',                    20    );
+add_filter( 'bbp_get_reply_content', 'wpautop',                            30    );
+
+// Run filters on topic content
+add_filter( 'bbp_get_topic_content', 'capital_P_dangit'                          );
+add_filter( 'bbp_get_topic_content', 'bbp_topic_content_append_revisions',  1, 2 );
+add_filter( 'bbp_get_topic_content', 'wptexturize',                         3    );
+add_filter( 'bbp_get_topic_content', 'convert_chars',                       5    );
+add_filter( 'bbp_get_topic_content', 'make_clickable',                      9    );
+add_filter( 'bbp_get_topic_content', 'force_balance_tags',                 25    );
+add_filter( 'bbp_get_topic_content', 'convert_smilies',                    20    );
+add_filter( 'bbp_get_topic_content', 'wpautop',                            30    );
+
 // Canonical
 add_filter( 'redirect_canonical', 'bbp_redirect_canonical' );

@@ -212,6 +258,8 @@
  *
  * This is used to clean-up any anonymous user data that is submitted via the
  * new topic and new reply forms.
+ *
+ * @uses add_filter() To add filters
  */
 function bbp_pre_anonymous_filters () {
 	// Post author name
Index: bbp-includes/bbp-reply-template.php
===================================================================
--- bbp-includes/bbp-reply-template.php	(revision 2779)
+++ bbp-includes/bbp-reply-template.php	(working copy)
@@ -306,24 +306,36 @@
  *
  * @since bbPress (r2553)
  *
- * @todo Have a parameter reply_id
- *
+ * @param int $reply_id Optional. reply id
  * @uses bbp_get_reply_content() To get the reply content
  */
-function bbp_reply_content() {
-	echo bbp_get_reply_content();
+function bbp_reply_content( $reply_id = 0 ) {
+	echo bbp_get_reply_content( $reply_id );
 }
 	/**
 	 * Return the content of the reply in the loop
 	 *
-	 * @since bbPress (r2553)
+	 * @since bbPress (r2780)
 	 *
+	 * @param int $reply_id Optional. reply id
+	 * @uses bbp_get_reply_id() To get the reply id
+	 * @uses post_password_required() To check if the reply requires pass
+	 * @uses get_the_password_form() To get the password form
+	 * @uses get_post_field() To get the content post field
 	 * @uses apply_filters() Calls 'bbp_get_reply_content' with the content
-	 * @uses get_the_content() To get the reply content
+	 *                        and reply id
 	 * @return string Content of the reply
 	 */
-	function bbp_get_reply_content() {
-		return apply_filters( 'bbp_get_reply_content', get_the_content() );
+	function bbp_get_reply_content( $reply_id = 0 ) {
+		$reply_id = bbp_get_reply_id( $reply_id );
+
+		// Check if password is required
+		if ( post_password_required( $reply_id ) )
+			return get_the_password_form();
+
+		$content = get_post_field( 'post_content', $reply_id );
+
+		return apply_filters( 'bbp_get_reply_content', $content, $reply_id );
 	}

 /**
Index: bbp-includes/bbp-topic-template.php
===================================================================
--- bbp-includes/bbp-topic-template.php	(revision 2779)
+++ bbp-includes/bbp-topic-template.php	(working copy)
@@ -334,6 +334,86 @@
 	}

 /**
+ * Output the content of the topic in the loop
+ *
+ * @since bbPress (r2780)
+ *
+ * @param int $topic_id Optional. Topic id
+ * @uses bbp_get_topic_content() To get the topic content
+ */
+function bbp_topic_content( $topic_id = 0 ) {
+	echo bbp_get_topic_content( $topic_id );
+}
+	/**
+	 * Return the content of the topic in the loop
+	 *
+	 * @since bbPress (r2780)
+	 *
+	 * @param int $topic_id Optional. Topic id
+	 * @uses bbp_get_topic_id() To get the topic id
+	 * @uses post_password_required() To check if the topic requires pass
+	 * @uses get_the_password_form() To get the password form
+	 * @uses get_post_field() To get the content post field
+	 * @uses apply_filters() Calls 'bbp_get_topic_content' with the content
+	 *                        and topic id
+	 * @return string Content of the topic
+	 */
+	function bbp_get_topic_content( $topic_id = 0 ) {
+		$topic_id = bbp_get_topic_id( $topic_id );
+
+		// Check if password is required
+		if ( post_password_required( $topic_id ) )
+			return get_the_password_form();
+
+		$content = get_post_field( 'post_content', $topic_id );
+
+		return apply_filters( 'bbp_get_topic_content', $content, $topic_id );
+	}
+
+/**
+ * Output the excerpt of the topic in the loop
+ *
+ * @since bbPress (r2780)
+ *
+ * @param int $topic_id Optional. Topic id
+ * @param int $length Optional. Length of the excerpt. Defaults to 100 letters
+ * @uses bbp_get_topic_excerpt() To get the topic excerpt
+ */
+function bbp_topic_excerpt( $topic_id = 0, $length = 100 ) {
+	echo bbp_get_topic_excerpt( $topic_id, $length );
+}
+	/**
+	 * Return the excerpt of the topic in the loop
+	 *
+	 * @since bbPress (r2780)
+	 *
+	 * @param int $topic_id Optional. topic id
+	 * @param int $length Optional. Length of the excerpt. Defaults to 100
+	 *                     letters
+	 * @uses bbp_get_topic_id() To get the topic id
+	 * @uses get_post_field() To get the excerpt
+	 * @uses bbp_get_topic_content() To get the topic content
+	 * @uses apply_filters() Calls 'bbp_get_topic_excerpt' with the excerpt,
+	 *                        topic id and length
+	 * @return string topic Excerpt
+	 */
+	function bbp_get_topic_excerpt( $topic_id = 0, $length = 100 ) {
+		$topic_id = bbp_get_topic_id( $topic_id );
+		$length   = (int) $length;
+		$excerpt  = get_post_field( $topic_id, 'post_excerpt' );
+
+		if ( empty( $excerpt ) )
+			$excerpt = bbp_get_topic_content( $topic_id );
+
+		if ( !empty( $length ) && strlen( $excerpt ) > $length ) {
+			$excerpt  = substr( $excerpt, 0, $length - 4 );
+			$excerpt .= '...';
+		}
+
+		return apply_filters( 'bbp_get_topic_excerpt', $excerpt, $topic_id, $length );
+	}
+
+/**
  * Output the status of the topic in the loop
  *
  * @since bbPress (r2667)
@@ -352,7 +432,8 @@
 	 * @param int $topic_id Optional. Topic id
 	 * @uses bbp_get_topic_id() To get the topic id
 	 * @uses get_post_status() To get the topic status
-	 * @uses apply_filters() Calls 'bbp_get_topic_status' with the
+	 * @uses apply_filters() Calls 'bbp_get_topic_status' with the status
+	 *                        and topic id
 	 * @return string Status of topic
 	 */
 	function bbp_get_topic_status( $topic_id = 0 ) {
Index: bbp-includes/bbp-users.php
===================================================================
--- bbp-includes/bbp-users.php	(revision 2779)
+++ bbp-includes/bbp-users.php	(working copy)
@@ -520,21 +520,16 @@
 /**
  * Get the total number of users on the forums
  *
- *  - Checks for a global $bbp_total_users, if it is set, then that is returned.
- *  - Runs the filter 'bbp_get_total_users', if we get anything other than false
- *     (strict check ===), then that is returned.
- *  - Runs its own query to count the users
- *
  * @since bbPress (r2769)
  *
  * @uses wp_cache_get() Check if query is in cache
- * @uses apply_filters() Calls 'bbp_get_total_users' with bool false
- * @uses wp_cache_set() Set the query in the cache
  * @uses wpdb::get_var() To execute our query and get the var back
+ * @uses wp_cache_set() Set the query in the cache
+ * @uses apply_filters() Calls 'bbp_get_total_users' with number of users
  * @return int Total number of users
  */
 function bbp_get_total_users() {
-	global $wpdb, $bbp_total_users;
+	global $wpdb;

 	if ( $bbp_total_users = wp_cache_get( 'bbp_total_users', 'bbpress' ) )
 		return $bbp_total_users;
Index: bbp-themes/bbp-twentyten/css/bbpress.css
===================================================================
--- bbp-themes/bbp-twentyten/css/bbpress.css	(revision 2779)
+++ bbp-themes/bbp-twentyten/css/bbpress.css	(working copy)
@@ -224,6 +224,10 @@
 	float: right;
 }

+p.form-allowed-tags {
+	width: 462px;
+}
+
 /* =Edit User
 -------------------------------------------------------------- */

Index: bbp-themes/bbp-twentyten/form-bbp_reply.php
===================================================================
--- bbp-themes/bbp-twentyten/form-bbp_reply.php	(revision 2779)
+++ bbp-themes/bbp-twentyten/form-bbp_reply.php	(working copy)
@@ -40,6 +40,10 @@
 							<textarea id="bbp_reply_content" tabindex="8" name="bbp_reply_content" cols="52" rows="6"><?php echo ( bbp_is_reply_edit() && !empty( $post->post_content ) ) ? $post->post_content : ''; ?></textarea>
 						</p>

+						<p class="form-allowed-tags">
+							<?php printf( __( '<label>You may use these <abbr title="HyperText Markup Language">HTML</abbr> tags and attributes:</label> %s','bbpress' ), '<code>' . bbp_allowed_tags() . '</code>' ); ?>
+						</p>
+
 						<?php if ( !bbp_is_reply_edit() ) : ?>
 							<p>
 								<label for="bbp_topic_tags"><?php _e( 'Tags:', 'bbpress' ); ?></label><br />
Index: bbp-themes/bbp-twentyten/form-bbp_topic.php
===================================================================
--- bbp-themes/bbp-twentyten/form-bbp_topic.php	(revision 2779)
+++ bbp-themes/bbp-twentyten/form-bbp_topic.php	(working copy)
@@ -55,6 +55,10 @@
 							<textarea id="bbp_topic_content" tabindex="10" name="bbp_topic_content" cols="52" rows="6"><?php echo ( bbp_is_topic_edit() && !empty( $post->post_content ) ) ? $post->post_content : ''; ?></textarea>
 						</p>

+						<p class="form-allowed-tags">
+							<?php printf( __( '<label>You may use these <abbr title="HyperText Markup Language">HTML</abbr> tags and attributes:</label> %s','bbpress' ), '<code>' . bbp_allowed_tags() . '</code>' ); ?>
+						</p>
+
 						<?php if ( !bbp_is_topic_edit() ) : ?>
 							<p>
 								<label for="bbp_topic_tags"><?php _e( 'Tags:', 'bbpress' ); ?></label><br />
Index: bbp-themes/bbp-twentyten/loop-bbp_replies.php
===================================================================
--- bbp-themes/bbp-twentyten/loop-bbp_replies.php	(revision 2779)
+++ bbp-themes/bbp-twentyten/loop-bbp_replies.php	(working copy)
@@ -54,7 +54,7 @@

 					<td class="bbp-reply-content">

-						<?php the_content(); ?>
+						<?php bbp_reply_content(); ?>

 					</td>

Index: bbp-themes/bbp-twentyten/single-bbp_reply.php
===================================================================
--- bbp-themes/bbp-twentyten/single-bbp_reply.php	(revision 2779)
+++ bbp-themes/bbp-twentyten/single-bbp_reply.php	(working copy)
@@ -58,7 +58,7 @@

 										<td class="bbp-reply-content">

-											<?php the_content(); ?>
+											<?php bbp_reply_content(); ?>

 										</td>

Index: bbp-themes/bbp-twentyten/single-bbp_topic.php
===================================================================
--- bbp-themes/bbp-twentyten/single-bbp_topic.php	(revision 2779)
+++ bbp-themes/bbp-twentyten/single-bbp_topic.php	(working copy)
@@ -71,7 +71,7 @@

 										<td class="bbp-topic-content">

-											<?php the_content(); ?>
+											<?php bbp_topic_content(); ?>

 										</td>
